### DISCUSSION TITLE
Remix examples?

### DISCUSSION THREAD

**Discussion by cgalpin** [lingui/js-lingui#1618] [Q&A]
Hi. Are there any examples of using Lingui with Remix?  I am not finding anything when i search.

tia,
charles

**Comment by timofei-iatsenko** [comment]
I've never used a Remix myself, so don't have any experience with it. If there is a way to load data from the server (and there is) and it uses react under the hood it should not be very different from nextjs integration. 

If you make a working example we would be very welcome to add it to the repo.

**Comment by cgalpin** [comment]
Ok, I'll give it a shot. But first I have a possibly stupid question - since neextjs uses swc, and remix uses esbuild, surely they need different aproaches?  In other words, will the swc-plugin work in remix? I don't think it uses babel

**Reply by timofei-iatsenko** [reply]
That might be the trickiest part. For sure we don't have a dedicated plugin for esbuild. But actually, esbuild don't allow any plugins which operates on AST. So if you want to apply plugins which are doing custom transformations, you need to use one of existing and supported transpilers together with esbuild. I would recommend to use swc because it's faster. I also don't know do remix allow somehow amending the esbuild configuration.  If they not, you probably will not be able to incorporate a macro to remix. 

But you still can use Lingui without Macro (and thus without any additional build plugins), macro is just a syntax sugar.

**Reply by timofei-iatsenko** [reply]
it seems it is not currently possible https://github.com/remix-run/remix/discussions/2278

**Reply by timofei-iatsenko** [reply]
Found this one, https://mortenbarklund.com/blog/styled-components-remix/ they mentioned `remix-esbuild-override` which helps with the overriding esbuild config

**Reply by timofei-iatsenko** [reply]
Than you can use https://github.com/sanyuan0704/esbuild-plugin-swc and enable any SWC plugins (styled-components, lingui, etc)

**Reply by cgalpin** [reply]
I appreciate the help, but I don't think I am going to be able to get this to work. I tried a number of things and even resorted to trying to get bard and bing to assist, but they are as clueless as I am :). From looking at that article, i think I'd need to to write a custom plugin like their styledComponentsPlugin but i don't know where to even begin. I don't have the time to spend - was just trying to evaluate remix andcover the major items like i18n (looking to switch an existing CRA based app).

It's a pity - i'd like to try remix, but not if I can't use lingui, even if just without macros!

thanks again

**Comment by andrii-bodnar** [comment]
Now it should be easier to integrate Remix with Lingui because it has [stable Vite support](https://remix.run/blog/remix-vite-stable) since version 2.7.0.

**Reply by andrii-bodnar** [reply]
Just tested it on the [remix-javascript](https://github.com/remix-run/remix/tree/main/templates/remix-javascript) template and at least the extraction works well using the Vite plugin.

**Reply by depsimon** [reply]
Hi @andrii-bodnar would you mind to share the result?

Did you just not used macro?

**Reply by andrii-bodnar** [reply]
Hi @depsimon, it looks like I didn't save the test project. As I recall, I set up the Lingui Vite plugin (https://lingui.dev/tutorials/setup-vite#setup-with-vitejs-plugin-react-swc) and used the macro in the code. The extraction worked fine in this case. It was a really simple test that probably didn't take into account all aspects of working with remix.

You can also check this - https://x.com/javiasilis/status/1774721034232406034

**Reply by depsimon** [reply]
Thanks, I ended up with a very similar result. Just wanted to make sure I was in the right track!

**Reply by andrii-bodnar** [reply]
@depsimon would you be interested in creating an [example](https://github.com/lingui/js-lingui/tree/main/examples) project?

**Reply by depsimon** [reply]
@andrii-bodnar Yes for sure. 

I'm looking for some time to gather all this up into an example project + a blog post. I might make a `remix-lingui` library as well that would make this process just as easy as with `remix-i18n`

**Reply by ericchernuka** [reply]
Which plugin for macros was everyone using? SWC or Babel? I'm only asking since Remix projects don't have the react plugin specified, but wondered if we run into issues declaring it?

**Reply by depsimon** [reply]
Using the babel plugin for macros.

Here's the gist of my vite.config

```ts
import { lingui } from "@lingui/vite-plugin";
import { vitePlugin as remix } from "@remix-run/dev";
import { defineConfig, loadEnv } from "vite";
import macrosPlugin from "vite-plugin-babel-macros";
import tsconfigPaths from "vite-tsconfig-paths";

export default defineConfig(async ({ mode }) => {
  return {
    plugins: [
      remix(),
      macrosPlugin(),
      lingui(),
      tsconfigPaths(),
    ],
  };
});
```

**Reply by ericchernuka** [reply]
Thanks @depsimon! One last question, do you do the load of the translations in entry.server or did you come up with a better place for the first render?

**Reply by depsimon** [reply]
I load the catalog in entry.server

**Reply by ericchernuka** [reply]
Interesting problem I'm running into is when I do the dynamic activate in the entry.server the root loader seems to have already been called so my translation in the loader returns the id on the first load. Then from then on, it's available. I oddly thought that the routes wouldn't have been executed yet and this would just work.

**Reply by depsimon** [reply]
I created a PR for the Remix Vite + Babel example I've been working on.

https://github.com/lingui/js-lingui/pull/1946

**Comment by miilord** [comment]
In Remix, server-side code (e.g., loader()) and client-side code are often written in the same file. Could this cause server-side translation messages to be unintentionally bundled into the client-side code?

**Reply by depsimon** [reply]
I suppose it can, especially if you load/change the catalog on the client-side.
